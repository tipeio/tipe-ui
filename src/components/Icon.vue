<template>
  <div
    :class="size"
    :data-tipe-ui="$options.name"
  />
</template>

<script>
import vueTypes from 'vue-types'

const sizes = ['small', 'medium', 'large']

export default {
  name: 'TipeIcon',
  props: {
    size: vueTypes.oneOf(sizes).def('medium')
  }
}
</script>

<style scoped>
[data-tipe-ui='TipeIcon'] {
  background-color: red;
}

[data-tipe-ui='TipeIcon'].small {
  height: 1rem;
  width: 1rem;
}

[data-tipe-ui='TipeIcon'].medium {
  height: 2rem;
  width: 2rem;
}

[data-tipe-ui='TipeIcon'].large {
  height: 3rem;
  width: 3rem;
}
</style>
