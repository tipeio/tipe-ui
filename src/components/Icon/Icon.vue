<template>
  <object
    :data="src"
    type="image/svg+xml">
    <img :src="src" >
  </object>
</template>

<script>
import vueTypes from 'vue-types'

export default {
  name: 'TipeIcon',
  props: {
    icon: vueTypes.string
  },
  computed: {
    src() {
      return require(`../../icons/${this.icon}.svg`)
    }
  }
}
</script>
