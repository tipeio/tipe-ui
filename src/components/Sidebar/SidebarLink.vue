<template>
  <a
    :class="rootClass"
    :data-tipe-ui="$options.name"
    :href="url"
    @mouseover="hovered = true"
    @mouseout="hovered = false"
  >
    <div class="icon-container">
      <tipe-icon
        :icon="icon"
        class="icon"
      />
    </div>
    <div
      v-if="!compressed"
      class="label">{{ label }}
    </div>
    <div
      :class="indicatorClass"
    />
  </a>
</template>

<script>
import vueTypes from 'vue-types'
import TipeIcon from '../Icon'

export default {
  name: 'TipeSidebarLink',
  components: {
    TipeIcon
  },
  props: {
    label: vueTypes.string,
    url: vueTypes.string.def('#'),
    icon: vueTypes.string,
    compressed: vueTypes.bool.def(false),
    active: vueTypes.bool.def(false)
  },
  data() {
    return {
      hovered: false
    }
  },
  computed: {
    isActive() {
      return this.active || this.hovered
    },
    rootClass() {
      return {
        compressed: this.compressed,
        active: this.isActive
      }
    },
    indicatorClass() {
      return {
        indicator: true,
        active: this.isActive
      }
    }
  }
}
</script>

<style lang="postcss">
[data-tipe-ui='TipeSidebarLink'] {
  display: flex;
  flex-direction: row;
  height: 5rem;
  background-color: var(--purple);
  align-items: center;
  justify-content: space-between;
  cursor: pointer;
  text-decoration: none;
}

[data-tipe-ui='TipeSidebarLink'].active {
  background-color: rgb(110, 92, 245);
}

.icon-container {
  flex: 0 0 4.6875rem;
  display: flex;
  height: 5rem;
  justify-content: center;
  align-items: center;
}

.icon {
  height: 1.5rem;
  width: 1.5rem;
}

.label {
  flex: 1 1;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  color: #fff;
}

.indicator {
  flex: 0 0 0.3125rem;
  height: 100%;
}

.indicator.active {
  background-color: #1be5b2;
}
</style>
