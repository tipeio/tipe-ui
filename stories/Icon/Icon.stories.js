import { storiesOf } from '@storybook/vue'

import TipeIcon from '../../src/components/Icon'

const style = () => ({
  height: '1rem',
  width: '1rem'
})

const icons = [
  'check',
  'document',
  'folder',
  'key',
  'number',
  'share',
  'switch',
  'user',
  'color',
  'edit',
  'gear',
  'link',
  'pin',
  'sign-out',
  'text',
  'view',
  'copy',
  'flow',
  'grid',
  'list',
  'search',
  'star-active',
  'trash',
  'wrench',
  'credit-card',
  'folder-open',
  'group',
  'markdown',
  'media',
  'send',
  'star',
  'user-circle',
  'options',
  'plus',
  'cross',
  'arrow-down',
  'arrow-up',
  'arrow-right',
  'arrow-left'
]

storiesOf('Icon', module).add('icons', () => ({
  components: { TipeIcon },
  computed: { style },
  template:
    '<div>' +
    icons.reduce((template, icon) => {
      return (
        template +
        `<tipe-icon :style="style" icon="${icon}" color="rgb(255,0,0)" />`
      )
    }, '') +
    '</div>'
}))
