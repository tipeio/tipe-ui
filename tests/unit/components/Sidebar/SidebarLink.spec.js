import { shallowMount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import SidebarLink from '@/components/Sidebar/SidebarLink.vue'
import Icon from '@/components/Icon'
import faker from 'faker'

describe('Sidebar', () => {
  describe('<template>', () => {
    it('matches previous snapshot', () => {
      const renderer = createRenderer()
      const wrapper = shallowMount(SidebarLink)

      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err)
        expect(str).toMatchSnapshot()
      })
    })

    it('<root> - has correct data-tipe-iu attibute', () => {
      const wrapper = shallowMount(SidebarLink)

      expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeSidebarLink')
    })

    it('<root> - should be a tag', () => {
      const wrapper = shallowMount(SidebarLink)

      expect(wrapper.is('a')).toBe(true)
    })

    it('<tipe-icon> - should exist', () => {
      const wrapper = shallowMount(SidebarLink)

      expect(wrapper.find(Icon).exists()).toBe(true)
    })
  })

  describe(':props', () => {
    it(':label - should match text', () => {
      const propsData = { label: faker.lorem.words(1) }
      const wrapper = shallowMount(SidebarLink, { propsData })

      expect(wrapper.text()).toBe(propsData.label)
    })

    it(':url - should default to #', () => {
      const wrapper = shallowMount(SidebarLink)

      expect(wrapper.attributes().href).toBe('#')
    })

    it(':url - should match href', () => {
      const propsData = { url: faker.internet.url() }
      const wrapper = shallowMount(SidebarLink, { propsData })

      expect(wrapper.attributes().href).toBe(propsData.url)
    })

    it(':icon - should be passed to Icon component', () => {
      const propsData = { icon: faker.lorem.word() }
      const wrapper = shallowMount(SidebarLink, { propsData })

      expect(wrapper.find(Icon).props().icon).toBe(propsData.icon)
    })

    it(':compressed(false) - should not add compressed class', () => {
      const wrapper = shallowMount(SidebarLink)

      expect(wrapper.classes()).not.toContain('compressed')
    })

    it(':compressed(true) - should add compressed class', () => {
      const propsData = { compressed: true }
      const wrapper = shallowMount(SidebarLink, { propsData })

      expect(wrapper.classes()).toContain('compressed')
    })
  })
})
