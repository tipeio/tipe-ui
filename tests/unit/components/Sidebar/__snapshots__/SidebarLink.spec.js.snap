// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar <template> matches previous snapshot 1`] = `
<a data-tipe-ui="TipeSidebarLink">
  <div class="icon-container">
    <TipeIcon-stub icon="" height="1.5rem" width="1.5rem" color="#FFF" __iconLoader__="function _default(icon) {
  return require(&quot;../../icons/&quot;.concat(icon, &quot;.svg&quot;)).default;
}" class="icon"></TipeIcon-stub>
  </div>
  <div class="indicator"></div>
</a>
`;
